packages/node-core/src/indexer/storeModelProvider/model/model.ts

```diff
@@ -0,0 +1,185 @@
+  ) {}
+
+  // Only perform bulk create data, without paying attention to any exception situations.
+  private async _bulkCreate(data: T[], tx?: Transaction): Promise<void> {
+    if (!data.length) return;
+
+    const uniqueMap = data.reduce(
+      (acc, curr) => {
+        acc[curr.id] = curr;
+        return acc;
+      },
+      {} as Record<string, T>
+    );
+    const waitCreateDatas = Object.values(uniqueMap);
+
+    await this.model.bulkCreate(waitCreateDatas as CreationAttributes<Model<T, T>>[], {
+      transaction: tx,
+      updateOnDuplicate: Object.keys(data[0]) as unknown as (keyof T)[],
+    });
+
+    if (tx) {
+      this.exporters.forEach((store: Exporter) => {
+        tx.afterCommit(async () => {
+          await store.export(data);
+        });
+      });
+    }
+  }
+
+  async get(id: string, tx?: Transaction): Promise<T | undefined> {
+    const record = await this.model.findOne({
+      // https://github.com/sequelize/sequelize/issues/15179
+      where: {id} as any,
+      transaction: tx,
+    });
+
+    return record?.toJSON();
+  }
+
+  async getByFields(filters: FieldsExpression<T>[], options: GetOptions<T>, tx?: Transaction): Promise<T[]> {
+    const fullOptions = getFullOptions(options);
+    // Query DB with all params
+    const records = await this.model.findAll({
+      where: {
+        [Op.and]: [...filters.map(([field, operator, value]) => ({[field]: {[operatorsMap[operator]]: value}}))] as any, // Types not working properly
+      },
+      limit: fullOptions.limit,
+      offset: fullOptions.offset,
+      order: [[fullOptions.orderBy as string, fullOptions.orderDirection]],
+      transaction: tx,
+    });
+
+    return records.map((r) => r.toJSON());
+  }
+
+  async set(id: string, data: T, blockHeight: number, tx?: Transaction): Promise<void> {
+    if (id !== data.id) {
+      throw new Error(`Id doesnt match with data`);
+    }
+
+    await this.bulkUpdate([data], blockHeight, undefined, tx);
+  }
+
+  async bulkCreate(data: T[], blockHeight: number, tx?: Transaction): Promise<void> {
+    await this.bulkUpdate(data, blockHeight, undefined, tx);
+  }
+
+  async bulkUpdate(data: T[], blockHeight: number, fields?: string[], tx?: Transaction): Promise<void> {
+    if (!data.length) return;
+
+    // TODO, understand why this happens, its also on the store cache
+    if (fields) {
+      throw new Error(`Currently not supported: update by fields`);
+    }
+
+    const uniqueMap = data.reduce(
+      (acc, curr) => {
+        acc[curr.id] = curr;
+        return acc;
+      },
+      {} as Record<string, T>
+    );
+    const uniqueDatas = Object.values(uniqueMap);
+
+    // Batch insert every 10000 data
+    const batchSize = 10000;
+    for (let i = 0; i < uniqueDatas.length; i += batchSize) {
+      const batchDatas = uniqueDatas.slice(i, i + batchSize);
+
+      if (!this.historical) {
+        await this._bulkCreate(batchDatas, tx);
+        continue;
+      }
+
+      await this.model.destroy({
```

Comment:
I set up a beforeBulkDestroy hook, which will only delete data at the current height.

Reply:
I don't think a hook is necessary for that, as there is no where else this could happen.

It also lacks any explanation, is the goal of this to remove all previous entries of the same entity in the same block?

Reply:
Yes, To ensure that only one record is retained at the current height. Case: Repeat the process of create - delete - create.